### DISCUSSION TITLE
Kong custom plugin with scheduler

### DISCUSSION THREAD

**Discussion by dinokey** [Kong/kong#10555] [Help] [answered]
Hello guys, 

I'm developing a kong custom plugin which required me to run scheduler inside the plugin.
Just need to initialize the scheduler once, but i still can not do that.

I use [`ngx.timer.every`](https://github.com/openresty/lua-nginx-module#ngxtimerevery) as scheduler.

I put my scheduler initialization code inside `handler.lua` in [`init_worker`](https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/) function. However, the scheduler intialized several times as `init_worker` function triggered several times. It is not what I want.

Is there any module or function that run once when plugin is initializing? so I can intialize scheduler at there.
Or maybe there is other way to do that?

**Comment by ghost** [comment, marked as answer]
> Just need to initialize the scheduler once, but i still can not do that.

Then you should not be using `ngx.timer.every`, but should use [ngx.timer.at](https://github.com/openresty/lua-nginx-module#ngxtimerat) which is a one-time timer.

> I put my scheduler initialization code inside `handler.lua` in [`init_worker`](https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/) function. However, the scheduler intialized several times as `init_worker` function triggered several times. It is not what I want.

Each worker process will execute all the code in `init_worker`, so if you only want one worker to execute the code, you can do it like this:

```lua
local ngx_worker_id = ngx.worker.id

    ...
    if ngx_worker_id() == 0 then
        ngx.timer.at(0, func)
    end
```

This way, the `func` will only be executed once during the Kong startup process.

**Reply by dinokey** [reply]
This is very helpful! Thanks @sabertobihwy

**Comment by esatterwhite** [comment]
Is there a way to get the plugin config, or the plugin id, so I can get the config from `init_worker` ?
